// surrealdb.js/esm/library/SurrealHTTP.js
import { NoConnectionDetails } from "../errors.js";
import { processUrl } from "./processUrl.js";
export class SurrealHTTP {
    constructor(url, { fetcher, } = {}) {
        Object.defineProperty(this, "url", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "authorization", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_namespace", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_database", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "fetch", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fetch = fetcher ?? fetch;
        this.url = processUrl(url, {
            ws: "http",
            wss: "https",
        });
    }
    ready() {
        return !!(this.url && this._namespace && this._database);
    }
    setTokenAuth(token) {
        this.authorization = `Bearer ${token}`;
    }
    createRootAuth(username, password) {
        this.authorization = `Basic ${btoa(`${username}:${password}`)}`;
    }
    clearAuth() {
        this.authorization = undefined;
    }
    use({ ns, db }) {
        if (ns)
            this._namespace = ns;
        if (db)
            this._database = db;
    }
    get namespace() {
        return this._namespace;
    }
    get database() {
        return this._database;
    }
    async request(path, options) {
        path = path.startsWith("/") ? path.slice(1) : path;
        if (!this.ready())
            throw new NoConnectionDetails();
        return (await this.fetch(`${this.url}/${path}`, {
            method: options?.method ?? "POST",
            headers: {
                "Content-Type": options?.plainBody
                    ? "text/plain"
                    : "application/json",
                "Accept": "application/json",
                "NS": this._namespace,
                "DB": this._database,
                ...(this.authorization
                    ? { "Authorization": this.authorization }
                    : {}),
            },
            body: typeof options?.body == "string"
                ? options?.body
                : JSON.stringify(options?.body),
        })).json();
    }
}
